### ISSUE TITLE
Throttle graphql queries during Janis builds

### ISSUE THREAD

**Issue by Niicck** [cityofaustin/techstack#3755]
### As a [Janis build] I want [to adjust how many simultaneous queries I send to Joplin during the build process] so that [queries don't get backed up and break Joplin].

**Is your feature request related to a problem? Please describe.**
/graphql queries timeout during the Janis build process.

**Describe the solution you'd like**
Make our Janis queries to Joplin smarter. Instead of sending so many simultaneous queries at once, throttle them so that only "x" are processed at a time. Once those "x" queries are resolved, then send over the next "x" queries. The value of "x" will be determined by experimentation. Maybe even set it as an environment variable.

It'll take some experimentation to know exactly what to do. But basically, we've got to stop doing what we're currently doing: sending too many queries all at once. The queries are getting backed up in Joplin, with response times increasing until they completely time out at 30 seconds.

**Comment by easherma** [comment]
Hey @Niicck look!
https://github.com/react-static/react-static/blob/master/docs/config.md#prefetchrate

Looks like we can probably use that to handle throttling, no?

**Comment by Niicck** [comment]
Very cool, let's give it a shot

**Comment by easherma** [comment]
We might not actually need this with https://github.com/cityofaustin/techstack/issues/3765

But either way for both, it would be good to have some test criteria that can be reproduced and measured.

**Comment by desigonz** [comment]
Ability to measure

**Comment by briaguya** [comment]
notes from testing:
### Things I've tried
* Making 1100+ location pages
  * When I did this, I was able to complete builds one at a time without any errors
* Running multiple builds at once (just a bunch of terminal tabs running build-full.sh)
  * This seems to cause build failures left and right
  * Even when I set the prefetchRate to 1 (which does slow down the build) it didn't lead to any fewer failures

### Conclusions
There isn't really an easy fix without a queueing system. It seems like this is a problem that will become quite a bit more manageable post publisher.

**Comment by briaguya** [comment]
@Niicck @easherma I decided to also test running 6 builds at a time against https://joplin-pr-async-and-pool.herokuapp.com/

There was still one failure but 5 out of 6 succeeded as opposed to building against staging where 5 out of 6 failed.

**Comment by briaguya** [comment]
@Niicck @easherma added both of you to this, not sure how to split it up

**Comment by Niicck** [comment]
@briaguya Do you have a branch that I can use to look at what you did? During your 1100+ location page test, what prefetch rate did you use? My Joplin PR is still 503ing even with the async changes merged in. So I want to use async + prefetch rate to see if that helps.

**Comment by Niicck** [comment]
Alright, using a prefetch rate=4 (b/c idk, we got 4 workers?, seemed like a safe guess), everything succeeded. It took 2-3 minutes than I would've liked, but it gets us somewhere.

Using a prefetch rate of 4 (in conjunction with async) significantly decreases our memory footprint. The 512MB peak is when I 503'd. I then ran 2 successful builds (one after the other) and barely made a dent in the memory usage. This tells me that we could definitely increase our prefetch rate higher than 4. We'll have to coordinate prefetch rate + gunicorn workers/connections + heroku hardware to get an optimal configuration.

<img width="1204" alt="Screen Shot 2020-03-01 at 12 12 12 PM" src="https://user-images.githubusercontent.com/14187277/75631169-2586fb00-5bb6-11ea-8c69-4684f53d1919.png">
